src/protocol/basics/sum_of_product/malicious.rs

```diff
@@ -70,30 +72,52 @@ where
     use crate::secret_sharing::replicated::malicious::ThisCodeIsAuthorizedToDowngradeFromMalicious;
 
     assert_eq!(a.len(), b.len());
+    let vec_len = a.len();
 
     let duplicate_multiply_ctx = ctx.narrow(&Step::DuplicateSop);
-    let random_constant_ctx = ctx.narrow(&Step::RandomnessForValidation);
-    let ax = a
-        .iter()
-        .map(|a| a.x().access_without_downgrade())
-        .collect::<Vec<_>>();
-    let arx = a.iter().map(|a| a.rx()).collect::<Vec<_>>();
-
-    let bx = b
-        .iter()
-        .map(|b| b.x().access_without_downgrade())
-        .collect::<Vec<_>>();
-
-    let (ab, rab) = try_join(
-        ctx.semi_honest_context()
-            .sum_of_products(record_id, ax.as_slice(), bx.as_slice()),
-        duplicate_multiply_ctx
-            .semi_honest_context()
-            .sum_of_products(record_id, arx.as_slice(), bx.as_slice()),
+
+    // generate shared randomness.
+    let prss = ctx.prss();
+    let duplicate_prss = duplicate_multiply_ctx.prss();
+    let (s0, s1): (F, F) = prss.generate_fields(record_id);
+    let (s0_m, s1_m): (F, F) = duplicate_prss.generate_fields(record_id);
+    let role = ctx.role();
+
+    // compute the value (d_i) we want to send to the right helper (i+1)
+    let mut right_sops: F = s1 - s0;
+    let mut right_sops_m: F = s1_m - s0_m;
+
+    for i in 0..vec_len {
+        let ax = a[i].x().access_without_downgrade();
+        let bx = b[i].x().access_without_downgrade();
+        let arx = a[i].rx();
+        right_sops += ax.right() * bx.right() + ax.left() * bx.right() + ax.right() * bx.left();
+        right_sops_m +=
+            arx.right() * bx.right() + arx.left() * bx.right() + arx.right() * bx.left();
+    }
+
+    // notify helper on the right that we've computed our value
+    let channel = ctx.mesh();
+    let channel_m = duplicate_multiply_ctx.mesh();
+    try_join(
+        channel.send(role.peer(Direction::Right), record_id, right_sops),
+        channel_m.send(role.peer(Direction::Right), record_id, right_sops_m),
```

Comment:
Note that this is pretty just all copy-pasted from src/protocol/basics/sum_of_product/semi_honest.rs

...I don't like that...

Reply:
yea we need to do something about it. I know that Richa is having the same issue, trying to implement a protocol that is *almost* the same for semi-honest and malicious. I added this to our Thursday agenda sync